backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIr.proto

```diff
@@ -1,79 +1,145 @@
+    optional bool is_enum_special = 11 [default = false];
 }
 
 message UniqId {
-    required int64 index = 1;
+    required uint64 index = 1;
+    required bool isLocal = 2;
 }
 
-// Inner descriptors used for inlining.
-message DeclarationDescriptor {
-    // These are ProtoBuf.Function, ProtoBuf.Class etc
-    oneof descriptor {
-        Function function = 1;
-        Class clazz = 2;
-        Property property = 3;
-        Constructor constructor = 4;
-    }
+message Coordinates {
+    required int32 start_offset = 1;
+    required int32 end_offset = 2;
 }
-message LocalDeclaration {
-    required DeclarationDescriptor descriptor = 1;
+
+/* ------ Top Level---------------------------------------------- */
+
+message IrDeclarationContainer {
+    repeated IrDeclaration declaration = 1;
 }
 
-message KotlinType {
-    required int32 index = 1;
-    required bool isCaptured = 2; // TODO: make it a flag set?
-    optional string debug_text = 3; // TODO: remove me
+message FileEntry { // TODO: extend me.
+    required string name = 1;
 }
 
-message Coordinates {
-    required int32 start_offset = 1;
-    required int32 end_offset = 2;
+message IrFile {
+    repeated UniqId declaration_id = 1;
+    required FileEntry file_entry = 2;
+    // TODO: we need a better string management. See metadata serialization as an example.
+    required string fq_name = 3;
+    //required IrDeclarationContainer declaration_container = 4;
+}
+
+message IrModule {
+    required string name = 1;
+    repeated IrFile file = 2;
+}
+
+/* ------ IrSymbols --------------------------------------------- */
+
+enum IrSymbolKind {
+    FUNCTION_SYMBOL = 1;
+    CONSTRUCTOR_SYMBOL = 2;
+    ENUM_ENTRY_SYMBOL = 3;
+    FIELD_SYMBOL = 4;
+    VALUE_PARAMETER_SYMBOL = 5;
+    RETURN_TARGET_SYMBOL = 6;
+    CLASS_SYMBOL = 7;
+    TYPE_PARAMETER_SYMBOL = 8;
+    VARIABLE_SYMBOL = 9;
+    ANONYMOUS_INIT_SYMBOL = 10;
+
+    STANDALONE_FIELD_SYMBOL = 11; // For fields without properties. WrappedFieldDescriptor, rather than WrappedPropertyDescriptor.
+    RECEIVER_PARAMETER_SYMBOL = 12; // ReceiverParameterDescriptor rather than ValueParameterDescriptor.
+}
+
+message IrSymbol {
+    required IrSymbolKind kind = 1;
+    required UniqId uniq_id = 2;
+    required UniqId top_level_uniq_id = 3;
+    optional string fqname = 4;
+    optional DescriptorReference descriptor_reference = 5;
+    //optional DescriptorReference top_level_descriptor_reference = 6;
+    //optional bool standalone_field = 5 [default = false];
+    //optional bool is_receiver = 6 [default = false];
+}
+
+/* ------ IrTypes --------------------------------------------- */
+
+enum IrTypeVariance { // Should we import metadata variance, or better stay separate?
+    IN = 0;
+    OUT = 1;
+    INV = 2;
+}
+
+message Annotations {
+    repeated IrCall annotation = 1;
+}
+
+message IrTypeBase {
+    required Annotations annotations = 1;
+    required IrTypeVariance variance = 2 [default = INV];
```

Comment:
Does `IrType` really has variance?

Reply:
Yes it does, https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/types/impl/IrTypeBase.kt#L17

Reply:
No, it is just an optimization trick to make `IrTypeBase` serve as both `IrType` and `IrTypeProjection`.

Reply:
Good catch. I've eliminated IrTypeBase from the proto description altogether, since we don't serialize KotlinType and what's left for IrType in there is just a list of annotations.